pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+locations of elixir and erlang code, index them an make them
+searchable. This is a large task that will not be accomplished in one
+commit, and this document will help to lay out the scope of the work,
+as well as collect thoughts and drive the implementation.
+
+Code intelligence will be oragnized into the following discrete areas:
+
+* **Discovery** - Finding points of interest in the project's codebase, dependencies and elixir core modules
+* **Caching** - Writing the discovered items to disk and ensuring the index is up to date with the state of the project at *all times*. The goal here is to never have to delete the `.lexical` directory.
+* **Indexing** - Storing the discovered artifacts in a local in-memory index that enable fast searching.
+* **Searching** ideally using something like the [sublime text's fuzzy search](https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/), but we definitely need some type of fuzzy search.
+* **Recognition** For several of the above examples (go to definition, find references), we need to be able to recognize what our cursor is hovering over. Is it a module? An atom? a variable? etc.
+
+# Discovery
+Lexical needs a flexible, expandable discovery engine. It needs to be
+able to examine both artifacts on disk like beam files and source
+code, and to index code that's being edited by the user. Right from
+the start, it needs to be able to find the following items:
+
+ * Modules
+ * Public and private module functions
+ * Macros
+ * Anonymous functions
+ * Structs
+ * Records
+ * Types
+ * Variables
+ * Atoms?
+ * Additional items, defined in plugins
+
+While investigating the above items, we will need to take into account
+the difference between usages and definitions. This will be pretty
+clear for most items (though aliasing and importing might muddy the
+waters somewhat), but atoms will be difficult, since there's no
+defintion to speak of, just a usage.
+
+# Caching
+> There are only two hard problems in computer science: Naming things, cache invalidation and off by one errors.
+
+Indexing the entire project, its dependencies, elixir and erlang when
+the language server starts is likely going to be too performance
+intensive to do (to wit, rewriting atoms during namespacing takes 15
+seconds on our moderately sized project), so we will need some sort of
+on-disk representation of the index.
+
+Such a representation will need to take the following into account:
+
+  * Branches can while the language server is running, invalidating large portions of the cache
+  * Project level search and replace can affect large numbers of files at once
+  * The versions of elixir and erlang can change via `asdf` whenever the user wants
+
+Basicallly, the project's assets can change at any time, and it's not
+guaranteed that the language server will be able to detect all of the
+changes, so the language server needs to treat this cache with some
+skepticism and constantly "check its work".
```

Comment:
would it be too crazy to have a CLI command that the user can run that tells the server to re_validate stuff? 

At least for the time being, an alternative to `rm -rf .lexical`

Not sure if a good idea, but I would enjoy not using the `rm`

Reply:
2. would we need to do like a dependency tree so we can isolate invalidations to _some_ part of the code, and avoid recompiling the whole thing? 

thinking about performance

Reply:
I'd rather lexical implement some kind of background check that would keep itself up to date.

Reply:
I guess it should not be too difficult since this is very standard for other tools, like compilers.
For nodejs I remember using this one to watch for files https://github.com/remy/nodemon
For unix there is `watch(1)`, but it sounds inefficient if you watch a whole directory: https://linux.die.net/man/1/watch
I guess it is only necessary to query the dates of file modification for existing files, and querying the directories for new files.

Reply:
The language server already has this: in [didChangeWatchedFiles](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles)